dlib/type_safe_union/type_safe_union_kernel_abstract.h

```diff
@@ -280,6 +283,33 @@ namespace dlib
                 equivalent to calling visit(std::forward<F>(f)) with void return type
         !*/
 
+        template <typename F>
+        void for_each(
+            F&& f
+        );
+        /*!
+            requires
+                - f is a callable object such that the following expression is valid for
+                  all types U in Types...:
+
+                    std::forward<F>(f)(const in_place_tag<U>& tag, type_safe_union<Types...>& x)
+            ensures
+                - the state of x is unchanged
```

Comment:
What does it do though? 

I don't totally get what this is for.  It's calling f() for each type in `Types...` right?  But why is that needed?  Like what does someone do with this?

Reply:
So deserialize() uses it

Reply:
The last test in the unit test shows a nice use case

Reply:
Remember initially you had reservations about supporting (de)serialization for ```std::variant``` because you argued users could do some metaprogramming like re-ordering the alternative types? In which case, serialization wouldn't work because the type indices would be all wrong? Well users could write (de)serialization code that uses ```for_each()``` and type identifiers such that serialization was invariant under re-ordering. This is just one example. In general, someone might want to iterate over the types. Like they might want a string description of all the types. Or they might want a hash code that hashes all the underlying types. There are loads of use cases.

Reply:
Eh, IDK.  So in the example you have in the unit test it's doing something that's not going to work.  Like check out https://en.cppreference.com/w/cpp/types/type_info/hash_code, those typeid hash codes are not stable even over reruns of the same binary.  There isn't any way in C++ to generate a stable ID like this for a type :(

If this is just about serialization stability over changes to the tuple it would be better to be able to say something like this: `type_safe_union<int, 1, foo_type, 2, other_type, 4>`.  That is, the user could set the id number for each type if they wanted.  That way adding or removing one isn't a huge deal since they can ensure the same ids always attach to the same types.

Reply:
Serialization was just one example. I just found `for_each()` useful. I get the feeling you don't. But `deserialize()` was essentially doing the same thing as `for_each()` so i thought i would write `for_each()` then `deserialize()` could use that. Then I found it useful for other stuff. If you don't want it then I can close the PR.

Reply:
With regards to generating hash codes for types, you can register them at runtime. So you're not relying on the type system to do this but at least it's stable since you're defining the codes manually.

Reply:
But the point of `for_each()` isn't just (de)serialization. It literally just does what it says on the tin and users, like me, find this useful.

Reply:
Sure, that's fine.  Can you make it be a global function though?  Like we shouldn't be adding methods to classes when they don't need access to the private details.  And call it `for_each_type()` to emphasize that it's looping over the types in a parameter pack.  Otherwise it's a little confusing since a type safe union only ever has one item in it.

Reply:
Yeah sure. I thought that was probably best to be honest. I feel like visit() should be global too but that's done now.